### PR TITLE
feat: adding tests for `FAST_ASSERT` environment

### PR DESCRIPTION
**Pull request by zepfred**
## Description of the change

This pull request includes tests that utilize the `FAST_ASSERT` and  `FULL_ASSERT` environments and tests for native builds. The objective is to identify any problems related to score corruption.

By default, the tests using the mentioned environment run for 30 seconds, and that was enough to find a bug in the Project Job Scheduling module.

A new job was created for spring-boot native build using the Liberica distribution.

### Development

- [NA] The changes have been covered with tests, if necessary.
- [NA] You have a green build, with the exception of the flaky tests.
- [NA] UI and JS files are fully tested, the user interface works for all modules affected by your changes (e.g., solve and analyze buttons).
- [NA] The network calls work for all modules affected by your changes (e.g., solving a problem).
- [NA] The console messages are validated for all modules affected by your changes.

### Code Review

- [x] This pull request includes an explanatory title and description.
- [ ] The GitHub issue is linked.
- [ ] At least one other engineer has approved the changes.
- [ ] After PR is merged, inform the reporter.

### CODE REVIEW

**Comment by triceo** [comment]
I think you overreached there a bit. :-) You tried to make "slowly" and "quickly" two different code paths, but it also resulted in twice the boilerplate. IMO it's enough if we just have the default and "slowly", where "slowly" runs everything that's default as well as everything that's slow.

Think of the users - they'll be copying this. They'll be trying to understand this. The less boilerplate you give them, the better.

File: .github/workflows/pull_request_long_running.yml [hunk 1]
```diff
@@ -77,6 +77,9 @@ jobs:
       - name: Quickly build timefold-solver
         working-directory: ./timefold-solver
         run: mvn -B -Dquickly clean install
-      - name: Test timefold-quickstarts in Native mode
+      - name: Test timefold-quickstarts (slowly tests)
```
Comment (triceo) [inline .github/workflows/pull_request_long_running.yml]:
```suggestion
      - name: Test timefold-quickstarts (slow tests)
```

File: hello-world/pom.xml [hunk 2]
```diff
@@ -1,137 +1,181 @@
...
          xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
-  <modelVersion>4.0.0</modelVersion>
+    <modelVersion>4.0.0</modelVersion>
 
-  <groupId>org.acme</groupId>
-  <artifactId>hello-world-school-timetabling</artifactId>
-  <version>1.0-SNAPSHOT</version>
+    <groupId>org.acme</groupId>
+    <artifactId>hello-world-school-timetabling</artifactId>
+    <version>1.0-SNAPSHOT</version>
```
Comment (triceo) [inline hello-world/pom.xml]: These changes are unrelated and unnecessary. Probably your IDE messed up?
Please remove from this file and others.

File: .github/workflows/pull_request_long_running.yml [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    branches: [stable, development, '*.x']
+    paths-ignore:
+      - 'LICENSE*'
+      - '.gitignore'
+      - '**.md'
+      - '**.adoc'
+      - '*.txt'
+
+jobs:
+#  build-quarkus:
```
Comment (triceo) [inline .github/workflows/pull_request_long_running.yml]: Please re-enable.

File: hello-world/src/test/java/org/acme/schooltimetabling/solver/TimetableFastAssertTest.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import ai.timefold.solver.core.config.solver.SolverConfig;
+
+import org.acme.schooltimetabling.TimetableApp;
+import org.acme.schooltimetabling.domain.Lesson;
+import org.acme.schooltimetabling.domain.Timetable;
+import org.junit.jupiter.api.Tag;
+import org.junit.jupiter.api.Test;
+
+@Tag("slowly")
+class TimetableFastAssertTest {
```
Comment (triceo) [inline hello-world/src/test/java/org/acme/schooltimetabling/solver/TimetableFastAssertTest.java]: So we are no longer doing `FULL_ASSERT`? I thought we'd be doing both.

File: technology/java-spring-boot/pom.xml [hunk 5]
```diff
@@ -154,8 +149,53 @@
...
+      <activation>
+        <activeByDefault>true</activeByDefault>
+      </activation>
+      <build>
+        <plugins>
+          <plugin>
+            <artifactId>maven-surefire-plugin</artifactId>
+            <version>${version.surefire.plugin}</version>
+            <configuration>
+              <excludedGroups>slowly</excludedGroups>
```
Comment (triceo) [inline technology/java-spring-boot/pom.xml]: By using the following, you'll be able to remove the entire `default` profile, possibly even both the profiles:

https://junit.org/junit5/docs/5.1.0/api/org/junit/jupiter/api/condition/EnabledIfSystemProperty.html

**Review by triceo — CHANGES_REQUESTED** [review]
I've left a few comments. Hopefully the Spring fail will be solved by the annotation I suggest; if not, please ping Chris, he's implemented the Spring native stuff.

File: hello-world/src/test/java/org/acme/schooltimetabling/solver/TimetableFastAssertTest.java (hunk 4 above)
Comment (zepfred) [reply hello-world/src/test/java/org/acme/schooltimetabling/solver/TimetableFastAssertTest.java]: Should we have tests for both environments?

File: .github/workflows/pull_request_long_running.yml (hunk 3 above)
Comment (zepfred) [reply .github/workflows/pull_request_long_running.yml]: Sure, I'm trying to fix the spring-boot issue first.

File: hello-world/src/test/java/org/acme/schooltimetabling/solver/TimetableFastAssertTest.java (hunk 4 above)
Comment (triceo) [reply hello-world/src/test/java/org/acme/schooltimetabling/solver/TimetableFastAssertTest.java]: I was expecting that.

File: technology/java-spring-boot/pom.xml (hunk 5 above)
Comment (zepfred) [reply technology/java-spring-boot/pom.xml]: Nice catch!

**Comment by zepfred** [comment]
> I've left a few comments. Hopefully the Spring fail will be solved by the annotation I suggest; if not, please ping Chris, he's implemented the Spring native stuff.

Your suggestion to remove the annotation `@Tag` worked perfectly. Thanks!

File: hello-world/src/test/java/org/acme/schooltimetabling/solver/TimetableEnvironmentTest.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+import org.junit.jupiter.api.condition.EnabledIfSystemProperty;
+
+@EnabledIfSystemProperty(named = "slowly", matches = "true")
+class TimetableEnvironmentTest {
+
+    @Test
+    void solve() {
+        solve(EnvironmentMode.FULL_ASSERT);
+        solve(EnvironmentMode.FAST_ASSERT);
+    }
```
Comment (triceo) [inline hello-world/src/test/java/org/acme/schooltimetabling/solver/TimetableEnvironmentTest.java]: I like this. Simple, no boilerplate, no profiles.
Do you mind propagating this to the rest of the assert tests?
(Spring Boot can do it differently, if this causes trouble.)
Maybe I'd make two methods out of this? So that you see in the log whether it was the full or the fast that failed.

File: hello-world/src/test/java/org/acme/schooltimetabling/solver/TimetableConstraintProviderTest.java [hunk 7]
```diff
@@ -1,14 +1,16 @@
...
 import ai.timefold.solver.test.api.score.stream.ConstraintVerifier;
+
 import org.acme.schooltimetabling.domain.Lesson;
 import org.acme.schooltimetabling.domain.Room;
 import org.acme.schooltimetabling.domain.Timeslot;
 import org.acme.schooltimetabling.domain.Timetable;
 import org.junit.jupiter.api.Test;
 
-import java.time.DayOfWeek;
-import java.time.LocalTime;
```
Comment (triceo) [inline hello-world/src/test/java/org/acme/schooltimetabling/solver/TimetableConstraintProviderTest.java:11]: Please reformat the imports using the order we use in the Solver.
(Alternatively: remove these unrelated changes, which will equal to the same result.)

**Review by triceo — APPROVED** [review]
Nice! I left one final comment; will merge when resolved.

File: hello-world/src/test/java/org/acme/schooltimetabling/solver/TimetableConstraintProviderTest.java (hunk 7 above)
Comment (zepfred) [reply hello-world/src/test/java/org/acme/schooltimetabling/solver/TimetableConstraintProviderTest.java:11]: I believe these changes match the configuration we use in the Solver. The `java.*` imports come first.

Comment (triceo) [reply hello-world/src/test/java/org/acme/schooltimetabling/solver/TimetableConstraintProviderTest.java:11]: Ah, in that case, I must remember it wrong. OK then.

**Comment by triceo** [comment]
There's test failures. :-(

**Comment by zepfred** [comment]
> There's test failures. :-(

Fixed!

File: hello-world/src/test/java/org/acme/schooltimetabling/solver/TimetableEnvironmentTest.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+                .withEnvironmentMode(environmentMode)
+                .withTerminationSpentLimit(Duration.ofSeconds(30)));
+
+        // Load the problem
+        Timetable problem = generateDemoData(TimetableApp.DemoData.SMALL);
+
+        // Solve the problem
+        Solver<Timetable> solver = solverFactory.buildSolver();
+        Timetable solution = solver.solve(problem);
+assertThat(solution.getScore()).isNotNull();    }
```
Comment (triceo) [inline hello-world/src/test/java/org/acme/schooltimetabling/solver/TimetableEnvironmentTest.java]:
```suggestion
        assertThat(solution.getScore()).isNotNull();    }
```

File: hello-world/src/test/java/org/acme/schooltimetabling/solver/TimetableEnvironmentTest.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+                .withEnvironmentMode(environmentMode)
+                .withTerminationSpentLimit(Duration.ofSeconds(30)));
+
+        // Load the problem
+        Timetable problem = generateDemoData(TimetableApp.DemoData.SMALL);
+
+        // Solve the problem
+        Solver<Timetable> solver = solverFactory.buildSolver();
+        Timetable solution = solver.solve(problem);
+        assertThat(solution.getScore()).isNotNull();    }
```
Comment (triceo) [inline hello-world/src/test/java/org/acme/schooltimetabling/solver/TimetableEnvironmentTest.java]:
```suggestion
        assertThat(solution.getScore()).isNotNull();
    }
```
